common/src/main/java/io/netty/util/internal/ObjectCleaner.java

```diff
@@ -98,20 +100,56 @@ public static void register(Object object, Runnable cleanupTask) {
 
         // Check if there is already a cleaner running.
         if (CLEANER_RUNNING.compareAndSet(false, true)) {
-            Thread cleanupThread = new FastThreadLocalThread(CLEANER_TASK);
-            cleanupThread.setPriority(Thread.MIN_PRIORITY);
+            objectCleanerThread = new FastThreadLocalThread(CLEANER_TASK);
+            objectCleanerThread.setPriority(Thread.MIN_PRIORITY);
             // Set to null to ensure we not create classloader leaks by holding a strong reference to the inherited
             // classloader.
             // See:
             // - https://github.com/netty/netty/issues/7290
             // - https://bugs.openjdk.java.net/browse/JDK-7008595
-            cleanupThread.setContextClassLoader(null);
-            cleanupThread.setName(CLEANER_THREAD_NAME);
+            objectCleanerThread.setContextClassLoader(null);
+            objectCleanerThread.setName(CLEANER_THREAD_NAME);
 
             // This Thread is not a daemon as it will die once all references to the registered Objects will go away
             // and its important to always invoke all cleanup tasks as these may free up direct memory etc.
-            cleanupThread.setDaemon(false);
-            cleanupThread.start();
+            objectCleanerThread.setDaemon(false);
+            objectCleanerThread.start();
+        }
+    }
+
+    /**
+     * Waits until the current cleaner thread has completed or the timeout has elapsed.
+     *
+     * Note: a new thread will be started if one is not running on a call to
+     * {@link #register(Object, Runnable)}, so this method is really only useful
+     * <strong>after</strong> your application has shut down and you want to wait for this
+     * thread to terminate.
+     *
+     * @return {@code true} if the cleaner thread has been terminated
+     */
+    public static boolean awaitInactivity(long timeout, TimeUnit unit) throws InterruptedException {
```

Comment:
This class is in an internal package and therefore not part of the Netty public api. The PR justification says it maybe useful to wait until all threads are complete. Can you provide more context here? The thread is currently not a deamon so the application should put terminate until the thread exits.

Reply:
I agree that my reasoning for this was not clear and more context is needed. In the elasticsearch tests, we shutdown a node, which includes stopping the Netty bootstrap(s). After this has been completed, our test framework checks for leaked threads and if there are some it waits for them, if the wait times out then it tries to interrupt all of the leaked threads, then waits again, and finally fails if the thread is still around.

Example (Netty 4.1.20.Final, JDK 9.04, Ubuntu 17.10):

```
  2> Jan 22, 2018 4:12:19 PM com.carrotsearch.randomizedtesting.ThreadLeakControl checkThreadLeaks
  2> WARNING: Will linger awaiting termination of 2 leaked thread(s).
  2> Jan 22, 2018 4:12:24 PM com.carrotsearch.randomizedtesting.ThreadLeakControl checkThreadLeaks
  2> SEVERE: 1 thread leaked from SUITE scope 
  2>    1) Thread[id=32, name=ObjectCleanerThread, state=TIMED_WAITING, group=TGRP-SSLClientAuthTests]
  2>         at java.base@9.0.4/java.lang.Object.wait(Native Method)
  2>         at java.base@9.0.4/java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:151)
  2>         at app//io.netty.util.internal.ObjectCleaner$1.run(ObjectCleaner.java:52)
  2>         at app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  2>         at java.base@9.0.4/java.lang.Thread.run(Thread.java:844)
  2> Jan 22, 2018 4:12:24 PM com.carrotsearch.randomizedtesting.ThreadLeakControl tryToInterruptAll
  2> INFO: Starting to interrupt leaked threads:
  2>    1) Thread[id=32, name=ObjectCleanerThread, state=TIMED_WAITING, group=TGRP-SSLClientAuthTests]
  2> Jan 22, 2018 4:12:27 PM com.carrotsearch.randomizedtesting.ThreadLeakControl tryToInterruptAll
  2> SEVERE: There are still zombie threads that couldn't be terminated:
  2>    1) Thread[id=32, name=ObjectCleanerThread, state=TIMED_WAITING, group=TGRP-SSLClientAuthTests]
  2>         at java.base@9.0.4/java.lang.Object.wait(Native Method)
  2>         at java.base@9.0.4/java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:151)
  2>         at app//io.netty.util.internal.ObjectCleaner$1.run(ObjectCleaner.java:52)
  2>         at app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
  2>         at java.base@9.0.4/java.lang.Thread.run(Thread.java:844)
```

In the same JVM more tests may be run after, which is why I'd like to see this thread stop. We can add a `ThreadLeakFilter` to allow this thread to "leak" and not fail but I don't want to pollute the next test.

I used ThreadDeathWatcher#awaitInactivity as precedent for this method but did not pay attention to the fact that the class is in the internal package.

I'm interested to hear what ideas you or @normanmaurer might have.